clients/tfchain-client-go/impl.go

```diff
@@ -37,145 +36,437 @@ type Versioned struct {
+	if conn := m.getExistingConn(); conn != nil {
+		return conn, nil
+	}
+
+	b := backoff.NewExponentialBackOff()
+	b.MaxInterval = 2 * time.Second
+	b.MaxElapsedTime = 4 * time.Second
+	b.InitialInterval = 500 * time.Millisecond
+	b.Multiplier = 2
+
+	var conn *poolConn
+	err := backoff.Retry(func() error {
+		// Check if we can get an existing connection
+		if c := m.getExistingConn(); c != nil {
+			conn = c
+			return nil
 		}
 
-		meta, err = cl.RPC.State.GetMetadataLatest()
-		if err != nil {
-			return errors.Wrapf(err, "error getting latest metadata at '%s'", endpoint)
+		m.mu.RLock()
+		poolSize := len(m.pool)
+		m.mu.RUnlock()
+
+		if poolSize >= m.config.MaxPoolSize {
+			return backoff.Permanent(ErrMaxPoolSizeReached)
 		}
 
-		t, err := getTime(cl, meta)
-		if err != nil {
-			return errors.Wrapf(err, "error getting node time at '%s'", endpoint)
+		select {
+		case m.checkChan <- struct{}{}:
+			log.Debug().Msg("triggered connection check")
+		default:
+			log.Debug().Msg("connection check already pending")
 		}
 
-		if time.Since(t) > acceptableDelay {
-			return fmt.Errorf("node '%s' is behind acceptable delay with timestamp '%s'", endpoint, t)
+		// time.Sleep(50 * time.Millisecond)
+		return ErrNoConnectionsAvailable
+	}, b)
+
+	return conn, err
+}
+
+func (m *manager) healthChecker() {
+	defer m.wg.Done()
+	// ticker := time.NewTicker(m.config.HealthCheckInterval)
+	// defer ticker.Stop()
+
+	for {
+		select {
+		case <-m.ctx.Done():
+			return
+		// case <-ticker.C:
+		// 	m.checkConnections()
+		case <-m.checkChan:
+			m.checkConnections()
 		}
+	}
+}
 
-		return nil
+func (m *manager) checkConnections() {
+	m.mu.Lock()
+	healthy := make([]*poolConn, 0, len(m.pool))
+	for _, conn := range m.pool {
+		if conn == nil {
+			continue
+		}
 
-	}, boff, func(err error, _ time.Duration) {
-		log.Error().Err(err).Msg("failed to connect to endpoint, retrying")
-	})
+		if !conn.isHealthy() {
+			log.Debug().Str("url", conn.url).Msg("closing unhealthy connection")
+			conn.close()
+			continue
+		}
+
+		// Check if connection is idle for too long if we have more than min pool size
+		if !conn.inUse.Load() && time.Since(time.Unix(conn.lastUsed.Load(), 0)) > m.config.MaxIdleTime && len(m.pool) > m.config.MinPoolSize {
+			log.Debug().Str("url", conn.url).Msg("closing idle connection")
+			conn.close()
+			continue
+		}
+
+		healthy = append(healthy, conn)
+	}
+
+	m.pool = healthy
+	m.mu.Unlock()
+	m.ensureMinConnections()
 
-	return cl, meta, err
 }
 
-// TODO: implement reusable connections instead of
-// closing the connection.
-func (p *mgrImpl) put(cl *Substrate) {
-	// naive put implementation for now
-	// we just immediately kill the connection
-	if cl.cl != nil {
-		cl.cl.Client.Close()
+func (m *manager) ensureMinConnections() {
+	log.Debug().Msg("ensuring minimum connections in the pool")
+	inUseCount := m.aquiredConnCount()
+	urls := shuffle(m.unusedURLs())
+	urls = append(urls, m.urls...)
```

Comment:
I don't get it here, isn't the urls returned from `unusedURLs` are already in m.urls? why not just use `m.urls`? and why shuffling them?

Reply:
When creating new connections, I want to prioritize unused URLs but won’t restrict attempts to them if all unused URLs are down. also, `unusedURLs()` can return 0 URLs.
Shuffling was implemented to distribute loads. Without shuffling, when an RPC node goes down, all nodes would attempt to connect to the same URL, causing spikes and high load.